src/main/webapp/signup.js

```diff
@@ -0,0 +1,162 @@
+/**
+ * @typedef SignUpData
+ * @property {string} gitHubAuthToken
+ * @property {string[]} interestTags
+ */
+import {gitHubAuthorizer} from './authorization.js';
+import {
+  standardizeFetchErrors,
+  makeRelativeUrlAbsolute,
+} from './js/fetch_handler.js';
+
+const AFTER_SIGNUP_REDIRECT = '/dashboard.html';
+
+/**
+ * Initializes the sign up form.
+ */
+function initSignupForm() {
+  const signupForm = document.getElementById('signup-form');
+  signupForm.addEventListener('submit', submitSignup);
+
+  signupForm.elements['github-link-button'].addEventListener(
+      'click', handleGitHubLink);
+  gitHubAuthorizer
+      .getFirebase().auth().onAuthStateChanged(handleAuthStateChanged);
+  // Handle auth state changed on page load in case of existing authentication.
+  handleAuthStateChanged(gitHubAuthorizer.getUser());
+}
+
+/**
+ * Handles the click event for the GitHub linking button.
+ * @param {Event} e The button click event.
+ */
+function handleGitHubLink(e) {
+  e.preventDefault();
+
+  gitHubAuthorizer.toggleSignIn().catch((error) => {
+    console.error(error);
+    alert('Encountered an error with GitHub authentication.'
+        + ' Please try again later.');
+  });
+}
+
+/**
+ * Updates the signup UI when the state of user GitHub authentication changes.
+ * @param {Firebase.User} user
+ */
+function handleAuthStateChanged(user) {
+  const signupForm = document.getElementById('signup-form');
+  const submitButton = signupForm.elements['submit-button'];
+  const gitHubLinkButton = signupForm.elements['github-link-button'];
+  if (user) {
+    submitButton.disabled = false;
+    gitHubLinkButton.innerText = 'Unlink Your GitHub Account';
+    gitHubLinkButton.classList.add('btn-success');
+    gitHubLinkButton.classList.remove('btn-purple');
+  } else {
+    submitButton.disabled = true;
+    gitHubLinkButton.innerText = 'Link Your GitHub Account';
+    gitHubLinkButton.classList.add('btn-purple');
+    gitHubLinkButton.classList.remove('btn-success');
```

Comment:
nit: consider renaming `btn-success` to something more descriptive of its functionality like`btn-unlink` or `btn-signout`. ditto for `btn-purple`.

Reply:
This is a solid suggestion, and it's directly in line with the [style guide](https://google.github.io/styleguide/htmlcssguide.html#ID_and_Class_Naming) too. However, `btn-success` is a class [tied to Bootstrap](https://getbootstrap.com/docs/4.0/components/buttons/#examples). We definitely could rename `btn-purple` though. Because this will affect more than my code, I'll discuss renaming this in our sync.